src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -0,0 +1,304 @@
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X PredX`
+  // (Where Y is a fresh variable)
+  Principal prin_y(FreshVar());
+
+  // This is `speaker says Y PredX`
+  Predicate generated_lhs = PushPrincipal("says_", prin_y, main_predicate);
+
+  Predicate y_can_act_as_x("canActAs",
+                           {prin_y.name(), attribute.principal().name()},
+                           Sign::kPositive);
+
+  Predicate speaker_says_y_can_act_as_x =
+      PushPrincipal("says_", speaker, y_can_act_as_x);
+
+  // This is `speaker says X PredX`
+  Predicate speaker_says_x_pred =
+      PushPrincipal("says_", speaker, main_predicate);
+
+  // This is the full generated rule:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X PredX`
+  return DLIRAssertion(
+      DLIRCondAssertion(generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                                        std::move(speaker_says_x_pred)}));
+}
+
+Predicate CanActAsToDLIR(const CanActAs& can_act_as) {
+  return Predicate(
+      "canActAs",
+      {can_act_as.left_principal().name(), can_act_as.right_principal().name()},
+      kPositive);
+}
+
+DLIRAssertion LoweringToDatalogPass::SpokenCanActAsToDLIR(
+    const Principal& speaker, const CanActAs& can_act_as) {
+  Predicate main_predicate = CanActAsToDLIR(can_act_as);
+
+  // "canActAs" facts are passed to principals via other canActAs facts in
+  // essentially the same way as attributes. This function adds extra
+  // rules to pass these around. If the `canActAs` under translation
+  // is `X canActAs Z`, then the rule we need to generate is:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X canActAsZ`
+  // (Where Y is a fresh variable)
+  Principal prin_y(FreshVar());
+
+  // This is `speaker says Y PredX`
+  Predicate generated_lhs = PushPrincipal("says_", prin_y, main_predicate);
+
+  Predicate y_can_act_as_x("canActAs",
+                           {prin_y.name(), can_act_as.left_principal().name()},
+                           Sign::kPositive);
+
+  Predicate speaker_says_y_can_act_as_x =
+      PushPrincipal("says_", speaker, y_can_act_as_x);
+
+  // This is `speaker says X canActAs Z`
+  Predicate speaker_says_x_can_act_as_z =
+      PushPrincipal("says_", speaker, main_predicate);
+
+  // This is the full generated rule:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X canActAs Z`
+  return DLIRAssertion(DLIRCondAssertion(
+      generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                      std::move(speaker_says_x_can_act_as_z)}));
+}
+
+std::pair<Predicate, std::vector<DLIRAssertion>>
+LoweringToDatalogPass::BaseFactToDLIRInner(const Principal& speaker,
+                                           const Predicate& predicate) {
+  std::vector<DLIRAssertion> pred_list = {};
+  return std::make_pair(predicate, std::move(pred_list));
```

Comment:
```suggestion
  return std::make_pair(predicate, {});

// if above does not work:  return std::make_pair(predicate, std::vector<DLIRAssertion>({}));
```

Reply:
It really did need to be the commented out one.